flips/2022-03-23-contract-imports-syntax.md

```diff
@@ -0,0 +1,153 @@
+switching betwen those environments hard.
+
+## Design Proposal
+
+### Declaring an Import
+
+There are currently two import replacement mechanisms used in Cadence contracts:
+
+#### **Imports referencing a placeholder**
+
+**Import syntax**
+
+`import <declarations to import> from 0x<placeholder>`
+
+Example:
+`import Foo from 0xFoo`
+
+**Mechanics**
+
+Source code is loaded and searched for `0x<placeholder>` string and then replaced with configured value. Configuration is not specified but it's assumed to be handled by the software loading the contract.
+
+**Used by**
+
+- FCL
+- Core contracts
+- Flow testing library
+
+#### **Imports referencing a relative path**
+
+**Import syntax**
+
+`import <declarations to import> from <relative path>` 
+
+Example:
+`import Foo from ./Foo.cdc`
+
+**Mechanics**
+
+Source code is loaded and parsed with the Cadence library, the import statements are iterated over and replaced with the values set in `flow.json` configuration. 
+
+**Used by**
+
+- VSCode Extension
+- CLI
+
+### Current Implementation Problems
+
+**Imports referencing a relative path**
+
+- there will not always be a contract present locally (or at least it shouldn't be as the network is the source of truth for a contract and downloading the contract has definite downsides especially contracts that are still able to update)
+- potential problems across different file systems
+
+**Imports referencing a placeholder**
+
+- Not a valid Cadence syntax
+- Impossible to know what contract is `0xFoo` really referencing by reading the contract code, the resolution of contract location is hidden in the complexity of the code
+
+### The Current State
+
+We can observe some common patterns established in the community:
+
+- Contracts are published on version control platforms (eg Github) and the best-established practice is for the code to be published on its own without other application code (which usually is part of another repository)
+- The community is using and switching between CLI and SDKs (FCL) and is facing problems between different import formats. This problem is validated with the existence of tools that handle switching between import formats (such as [github link])
+
+### Possible Solutions?
+
+**New Import Syntax**
+
+`import <declarations to import> from "<namespaced location>"`
+
+Example:
+
+`import NonFungibleToken from "onflow/NonFungibleToken"`
+
+The source files should be accompanied by a configuration/dependency resolution file, defining locations of contracts addresses on a specific network. Including this file is beneficial so developers exploring contract source files in public repositories can understand which contracts this Cadence code imports. Think of this as go.mod, package.json or any other implementation of this idea where source code that imports something needs to define what that is in the supporting "configuration" file.
+
+Example flow.json for the above syntax:
```

Comment:
As discussed, maybe point out this is only an example how a tool like the CLI may use the syntax proposed this standard, and the behaviour of the tools is not part of this proposal